codalab/server/bundle_manager.py

```diff
@@ -310,9 +310,12 @@ def _bring_offline_stuck_running_bundles(self, workers):
 
     def _schedule_run_bundles_on_workers(self, workers, staged_bundles_to_run, user_info_cache):
         """
-        Schedules STAGED bundles to run on the given workers. Always tries to schedule bundles to
-        run on workers that are owned by the owner of each bundle first. If there are no such qualified
-        private workers, uses CodaLab-owned workers, which have user ID root_user_id.
+        Schedule STAGED bundles to run on available workers based on the following logic:
+        1. If the bundle requests to run on a specific worker, tries to schedule the bundle
+           to run on a worker that has a tag exactly match with bundles' request_queue.
+        2. If the bundle doesn't request to run on a specific worker,
+          (1) try to schedule the bundle to run on a private worker that is owned by the bundle's owner
+          (2) if there is no such qualified private worker, uses CodaLab-owned workers, which have user ID root_user_id.
         :param workers: a WorkerInfoAccessor object containing worker related information e.g. running uuid.
```

Comment:
@percyliang and @bkgoksel Logic changes are reflected in the above comments from line 313 to line 318. No rerouting supported in the future dispatching logic anymore, meaning if there is no matched private worker can be found, the program won't try to reroute the bundle to run on any other available workers. The program will just do a **exact match** between **bundle.metadata.request_queue** and **worker's tag**